### DISCUSSION TITLE
cPanel and Passenger

### DISCUSSION THREAD

**Discussion by mcqueena9845** [directus/directus#3394] [Archive]
I'm new to Directus and Nodejs.  I'm trying to get Directus running on cPanel with CentOS 7.  Nodejs is up and working.  I can create and register a sample app like this: https://docs.cpanel.net/knowledge-base/web-services/how-to-install-a-node.js-application/.  I'm able to do the manual install of directus with npm, and register it with the Application Manager.  https://docs.cpanel.net/cpanel/software/application-manager/  I added the Environment Variables in the Application Manager. The Application Manager is able to verify the dependencies for Directus.  My understanding is that this uses Phusion Passenger.  Passenber is looking for an app.js file in my application root.  What do I put in the app.js file to start Directus with Passenger?

**Comment by mcqueena9845** [comment]
I gave up on this and ended up with a standalone server.

**Comment by Dominic-Marcelino** [comment]
If someone else is looking for this:

I’ve never used cPanel, but it should be the same as with plesk and Phusion. All you need is creating a start-file and an empty folder for the root directory in the main directory like this one

```js
const { default: logger } = require('directus/dist/logger');
const { default: start } = require('directus/dist/start');

start().catch(function(err) {
    logger.error("Error", { err });
});
```

Source: [Official Discord] (https://discord.com/channels/725371605378924594/741330910594465823/789131069962059793)

**Reply by lahdekorpi** [reply]
Just wanting to confirm that this + setting PORT env as passenger did indeed work with cPanel.

**Reply by OzzyTheGiant** [reply]
@Dominic-Marcelino Would like some help on this. I've been using the script above since rc-62 but now that I upgraded to 9.4.2, it doesn't work anymore. Keep getting errors that the node modules don't exist, saying it "cannot find module "/path/to/node_modules/directus/dist/dist/logger". Weirdly it's appending an extra /dist folder when I didn't tell it to do so. Removing "dist" from the require statmement doesn't help either. Tried relative imports and nothing either.

**UPDATE** - So I figured out through the source code that I should just `require("directus/server")`. I still don't understand why `require` is adding that extra dist folder to the path, but that's how I got it to work on localhost. Then on the production server, it still doesnt' work. There are no errors other than some DeprecationWarning for "GLOBAL" and when I try to load the admin panel, Passenger still spits out a 500 error. Running the app.js file directly using `node app.js` mentions Illegal Instruction. Idk if this app simply requires certain system permissions, but if that's the case then this is a real disappointment as I'm running on a shared server and I can't upgrade that hosting account at all.

**Reply by rijkvanzanten** [reply]
> that I should just require("directus/server"). I still don't understand why require is adding that extra dist folder to the path

It's because we started relying on Node's `export` field in package.json to explicitly tell the module system what files to include on require (Node's transition from CJS to ESM is a pain point and discussion I won't get into here..)

As for the Illegal Instruction, that sounds like something more system native than anything we could throw in Node itself 🤔 

How are you installing the Node Modules? It could be that a system native dependency was built against the wrong target, which sometimes happens when you `npm install` outside of the VPS, and then copy/ftp it in.

**Reply by OzzyTheGiant** [reply]
@rijkvanzanten Agreed, after going through this a few more times, it pretty much seems like the Illegal Instruction is something related to Dreamhost not allowing idle processes to run on Shared Servers (not that I was planning to do that, just wanted to test that Directus was working without Passenger first).

And yes, I thought so too; I first did `npm install` to upgrade the packages and then after the errors, I cleaned out node_modules and did a brand new `npm install`; it did not work either of those times. On both times, I also removed package-lock.json just to be sure. Checked around and noticed the server is Ubuntu 18.04, not sure if that affects this (I doubt it).

Additional factors that may be at play are different Node.js versions - My localhost uses 14.18 and production server uses 12.21 via NVM; I couldn't even install v14.

I tried following more or less the guide for Shared Hosting via Plesk. No, Plesk doesn't exist on Dreamhost but I used the app.js script mentioned in that guide. It SEEMS to work (meaning there's no errors in the logs other than the DeprecationWarning for GLOBAL), but Passenger continues to spit out error 500, so I'm kinda just debugging in the dark here with all of these silent errors.

**Reply by rijkvanzanten** [reply]
Hmm yeah hard to tell then.. The vague error message `Illegal Instruction` does indicate that there's some issue in the exact running environment (outside of Directus' JS code itself) that needs to be resolved. `npm install` does sometimes download pre-built binaries based on the current environment. It might be worth a shot compiling everything fresh in the correct environment to ensure the problem doesn't exist within a native extension, this should be possible through the `npm rebuild` command

**Reply by OzzyTheGiant** [reply]
Tried rebuild, looked at all the lines and everything seemed normal, no errors, just one warning about a sqlite library but nothing major. Still getting the same 500 error and no error.log output. About to call it quits as I think the Dreamhost shared server is just not budging on this. It really is unfortunate that it all broke from rc-62 upgrading to 9.4. I think I might just have to host on a VPS instead and use a PHP script for the contact form in the meantime.

**Reply by rijkvanzanten** [reply]
Hmm yeah that's super annoying 😑 I do recall having seem some similar issues around GLIB_C versions (used for `argon2`) in the past, but without more information logged from Dreamhost, it'll be impossible to confirm / rule anything out

**Reply by turbolaserhawk** [reply]
@lahdekorpi how did you find Directus 9 performance on cPanel?  I am interested in trying it for a multi-tenant Directus setup.  The built in DNS, analytics, firewall, etc all appeal to me, and I am a long time cPanel user since about 2004.  The thing I am concerned about is that it won't scale as nicely as LAMP stack on cPanel.  What was your experience like?

**Comment by magosis** [comment]
As of the changeover to node 18 as a requirement @Dominic-Marcelino 's solution does not seem to work on WHM/passenger does anyone have the new needed entrypoint script for this?
